File: leetcode/Algorithms/542.01Matrix/Solution.py
from collections import deque


class Solution(object):
    magic = 250
    height = 0
    length = 0

    def updateMatrix(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[List[int]]
        """
        if len(matrix) == 0:
            return matrix

        self.height, self.length = len(matrix), len(matrix[0])
        for i in range(self.height):
            for j in range(self.length):
                if matrix[i][j] != 0:
                    matrix[i][j] = self.magic

        queue = deque([])
        for i in range(self.height):
            for j in range(self.length):
                if matrix[i][j] == 0:
                    self.BFS(matrix, queue, i, j, 1)

        while len(queue) != 0:
            point = queue.popleft()
            self.BFS(matrix, queue, point[0], point[1], matrix[point[0]][point[1]] + 1)

        return matrix

    def BFS(self, matrix, queue, i, j, weight):
        if i > 0 and matrix[i - 1][j] == self.magic:
            matrix[i - 1][j] = weight
            queue.append([i - 1, j])
        if j > 0 and matrix[i][j - 1] == self.magic:
            matrix[i][j - 1] = weight
            queue.append([i, j - 1])
        if i + 1 < self.height and matrix[i + 1][j] == self.magic:
            matrix[i + 1][j] = weight
            queue.append([i + 1, j])
        if j + 1 < self.length and matrix[i][j + 1] == self.magic:
            matrix[i][j + 1] = weight
            queue.append([i, j + 1])

File: leetcode/Algorithms/542.01Matrix/Solution.java
public class Solution {
    public int[][] updateMatrix(int[][] matrix) {
        if (matrix.length == 0) return matrix;
        int height = matrix.length, length = matrix[0].length, MAGIC = 250;

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < length; j++) {
                if (matrix[i][j] != 0) matrix[i][j] = MAGIC;
            }
        }

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < length; j++) {
                if (matrix[i][j] == 0) DFS(matrix, i, j, 0);
            }
        }

        return matrix;
    }

    private void DFS(int[][] matrix, int i, int j, int weight) {
        matrix[i][j] = weight++;
        if (i > 0 && matrix[i - 1][j] > weight) DFS(matrix, i - 1, j, weight);
        if (j > 0 && matrix[i][j - 1] > weight) DFS(matrix, i, j - 1, weight);
        if (i + 1 < matrix.length && matrix[i + 1][j] > weight) DFS(matrix, i + 1, j, weight);
        if (j + 1 < matrix[0].length && matrix[i][j + 1] > weight) DFS(matrix, i, j + 1, weight);
    }
}
